clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -470,42 +473,85 @@ class MarkDeviceFunction : public RecursiveASTVisitor<MarkDeviceFunction> {
 
   // The call graph for this translation unit.
   CallGraph SYCLCG;
-  // The set of functions called by a kernel function.
-  llvm::SmallPtrSet<FunctionDecl *, 10> KernelSet;
-  // The set of recursive functions identified while building the
-  // kernel set, this is used for error diagnostics.
+  // Record the mapping between each SYCL kernel or SYCL_EXTERNAL function and
+  // functions called by it.
+  llvm::DenseMap<FunctionDecl *, llvm::SmallPtrSet<FunctionDecl *, 10>>
```

Comment:
This seems to be re-implementing a CallGraphNode.  Is there a reason we need to re-store these?

Reply:
Hi, @erichkeane 
I didn't quite understand the question here, did you ask why the patch dropped the "KernelSet" and introduced the "SYCLKernelInvokeMap" here?
Thanks very much.

Reply:
I'm basically asking 'why' on a lot of things throughout this patch.  I'm unconvinced that KernelSet was necessary, and I'm even further unconvinced that SYCLKernelInvokeMap is required.  I think this patch is well-intentioned but is the straw that breaks the camels back here.

The complexity of this type and section of code is too much after this patch.  We've hit 'unmaintainable' unfortunately.  So if you can do as @AaronBallman suggested and put together a bunch of lit tests that validate the edge-cases that you found, he and I will work on refactoring this code ensuring it fixes the issues you saw.